### ISSUE TITLE
Automatically reconnect if a timeout occurs

### ISSUE THREAD

**Issue by Chronophylos** [museun/twitchchat#119]
Add a option to allow reconnecting to chat when a timeout occurs

**Comment by museun** [comment]
I'll look into it, the current design of the upcoming versions will actually remove the `TcpStream` from the crate so it can be used with async-std (and smol). There will be indv. `connect`/`connect_tls` functions under runtime modules (`tokio::connect`, `async_std::connect`) that are enabled via a feature flag.

For this to work with a 'restartable' loop, I'd need to have the user provide a connect function that can be called each iteration. Also, a backoff procedure should be provided. I'm not sure the rules for Twitch and whether it'd be best to use some exponential backoff or what.

so something like
```rust
async fn run<C, B, RW, CE>(
    dispatcher: Dispatcher, 
    rate_limit: RateLimiter, 
    connect_func: C, 
    backoff_func: B
) -> Result<Status, Error> 
where 
    C: Result<RW, CE>,
    RW: Future<Item = AsyncRead+AsyncWrite+Unpin+'static>,
    CE: std::error::Error,
    B: Fn() -> std::time::Duration,
{
    todo!()
}
```

and then I can provide some function wrappers (`tokio::connect()`, `async_std::connect()`, `tokio::connect_and_retry()`)

(names/types totally pending and this is 'off the cuff')

Currently, a simple function that calls connect/run and returns the future could work. I'll write one up in a bit to show off how that'd look. But the runner taking ownership of itself (via `mut self`) makes it a bit more messy than I'd like.

I apologize for the delay, unusually busy weekend.

**Comment by Chronophylos** [comment]
Thank you for your answer.

I checked the twitch docs and they suggest a exponential backoff. https://dev.twitch.tv/docs/irc/guide#re-connecting-to-twitch-irc

> Currently, a simple function that calls connect/run and returns the future could work. I'll write one up in a bit to show off how that'd look. But the runner taking ownership of itself (via `mut self`) makes it a bit more messy than I'd like.

It would help me a lot if you could provide such a function.

**Comment by museun** [comment]
This is a bit more than a function, but this is a runnable example showing a way of doing it with the v0.10 release
```rust
use tokio::stream::StreamExt as _;

use twitchchat::{events, Control, Dispatcher, Error, RateLimit, Runner, Status, UserConfig};

async fn connect_loop(
    user_config: &UserConfig,
    dispatcher: Dispatcher,
    channels: &[String],
) -> Result<(), Error> {
    // exponential backoff in seconds
    // starting at 10, so 10 -> 20 -> 40 -> ...
    fn make_backoff() -> impl Iterator<Item = std::time::Duration> {
        std::iter::successors(Some(10), |d| Some(d * 2)).map(std::time::Duration::from_secs)
    }

    // make the backoff sequence
    let mut backoff = make_backoff();

    loop {
        // clone the dispatcher for each iteration (its moved into the 'await' yield point below)
        let dispatcher = dispatcher.clone();

        // try to connect to twitch with the provided user_config
        match twitchchat::connect(&user_config).await {
            Ok(conn) => {
                // we connected, so start the run loop
                let (runner, control) = Runner::new(dispatcher.clone(), RateLimit::default());

                // reset the backoff because we've successfully completed a connection
                backoff = make_backoff();

                // start the run loop
                let fut = run_loop(control.clone(), dispatcher, channels);

                // wait for the end of the 'run_loop' or the end of the main internal loop
                tokio::select! {
                    should_we_quit = fut => {
                        if should_we_quit {
                            control.stop()
                        }
                    }
                    status = runner.run(conn) => {
                        match status {
                            // stop the loop if the user canceled it
                            Ok(Status::Canceled) => { return Ok(()) }
                            Ok(Status::Eof) | Ok(Status::Timeout) => {
                                /* try to reconnect with the next delay */
                            }
                            Err(err) => {
                                /* try to reconnect with the next delay */
                                eprintln!("got an error while running: {}", err);
                            }
                        }
                    }
                }
            }

            // we got an error, so try to reconnect
            Err(err) => eprintln!("cannot connect: {}", err),
        };

        // get the next exp. sequence number
        let secs = backoff.next().unwrap();
        eprintln!("waiting for: {:?}", secs);
        // and wait for that many seconds
        tokio::time::delay_for(secs).await;
    }
}

async fn run_loop(mut control: Control, mut dispatcher: Dispatcher, channels: &[String]) -> bool {
    let mut join = dispatcher.subscribe::<events::Join>();
    let mut part = dispatcher.subscribe::<events::Part>();
    let mut pm = dispatcher.subscribe::<events::Privmsg>();

    wait_and_join(&mut control, &mut dispatcher, channels).await;

    let mut writer = control.writer().clone();
    let just_for_show = async move {
        eprintln!("we're quitting in 3 seconds..");
        tokio::time::delay_for(std::time::Duration::from_secs(3)).await;
        writer.raw("/quit").await.unwrap()
    };

    // pin the future
    tokio::pin! { just_for_show };

    loop {
        tokio::select! {
            Some(msg) = join.next() => {
                eprintln!("{} joined {}", msg.name, msg.channel);
            }
            Some(msg) = part.next() => {
                eprintln!("{} left {}", msg.name, msg.channel);
            }
            Some(msg) = pm.next() => {
                eprintln!("[{}] {}: {}", msg.channel, msg.name, msg.data);
            }

            _ = &mut just_for_show => {
                eprintln!("quitting manually");
                return false;
            }
            else => { break }
        }
    }

    true
}

async fn wait_and_join(control: &mut Control, dispatcher: &mut Dispatcher, channels: &[String]) {
    let ready = dispatcher.wait_for::<events::IrcReady>().await.unwrap();
    eprintln!("our name: {}", ready.nickname);

    let w = control.writer();
    for channel in channels {
        eprintln!("joining: {}", channel);
        let _ = w.join(channel).await;
        eprintln!("joined");
    }
    eprintln!("joined all channels")
}

#[tokio::main]
async fn main() {
    let channels = std::env::var("TWITCH_CHANNELS")
        .map(|s| s.split(",").map(ToOwned::to_owned).collect::<Vec<_>>())
        .unwrap();

    let (nick, pass) = twitchchat::ANONYMOUS_LOGIN;
    let user_config = twitchchat::UserConfig::builder()
        .name(nick)
        .token(pass)
        .capabilities(&[
            twitchchat::Capability::ChatRooms,
            twitchchat::Capability::Commands,
            twitchchat::Capability::Membership,
            twitchchat::Capability::Tags,
        ])
        .build()
        .unwrap();

    let dispatcher = Dispatcher::new();
    // you can set up stuff with the dispatch here, or in the `run_loop`

    connect_loop(&user_config, dispatcher, &channels)
        .await
        .unwrap();
}
```

**Comment by museun** [comment]
The important bit the `fn connect_loop`.

We make an infinite sequence of `[10*2..]`.

Then we have an infinite loop. In this loop we:
1) connect using the provided user config
    1) reset our backoff sequence on successful connects (so the next one starts back at `10`).
    1) make the main `future` that does whatever logic.
    1) check to see if that future returned `true`/`false` OR whether the twitchchat event loop returned.
        1) If the user `future` returned `true`, we end the loop (this is just to simulate / show a 2nd way of stopping). 
            1) otherwise, go to the backoff section.
        1) If we got an `Error` / `Eof` / `Timeout`, we go to the backoff section
            1) otherwise we return from the loop. The 'Canceled' status can/will happen when Control::stop() is called.
1) If we get a connection error, try again.

and finally, get the next sequence and 'block' the task for that long (e.g. `10->20->40`. until `make_backoff` resets it)

Note, I don't tend to turn the crate into a bot one, it just provides an abstract way of creating bots/clients. Putting a loop like this in the crate will make it awkward for people (like me) who need more control over the before/after parts of the connection/run sequence.

So, when I do add it, it'll be a separate function that just does something similar via callback/provided futures (e.g. `run` and `run_with_retry`).

In the future, I may plan on writing a higher level crate (`twitchchat_bot` or something) that acts a bot framework. But I have no interest/motivation of doing it at this moment.

**Comment by Emilgardis** [comment]
I don't think this makes sense for this crate, imo reconnect should be a client detail.

Edit: actually, maybe a `run_with_retry` could work nixely, where you perhaps supply an async `Fn()->bool` where `true` indicates retry.

**Comment by museun** [comment]
Something like this could work

```rust
enum Retry {
    Delay(Duration),
    Immediately,
    Cancel
}
Fn(Status) -> impl Future<Output = Retry>
```

Edit: But if I move the delay into the user code, I wouldn't have to rely on a timer.

This allows the user to block if they want to or not.
```rust
Fn(Status) -> impl Future<Output = bool>
// and
async move { delay_for(10.sec()).await; true }
```

**Comment by museun** [comment]
I've implemented a preliminary version of this in #129
